File: Algorithms/Sorting/Counting-Sort/C++/counting_sort.cpp
#include <iostream>
using namespace std;

void counting_sort(int *arr, const int size, const int MAX_VALUE=10){
  int counts[MAX_VALUE] = { 0 };

  int i, j;
  for(i = 0; i < size; i++){
    counts[arr[i]]++;
  }
  
  for(i = j = 0; i < MAX_VALUE; i++){
    while(counts[i]--){
      arr[j++] = i;
    }
  }
}

int main(){
  int arr[] = {1,4,1,2,7,5,2};
  counting_sort(arr,7);

  for(int i = 0; i < 7; i++){
    cout << arr[i] << " ";
  }
  
  return 0;
}

File: Algorithms/Sorting/Counting-Sort/Python/counting_sort.py

def counting_sort(arr, MAX_VALUE=10):
    # 0. Initialize counts array with zeroes:
    cnt = [0] * MAX_VALUE
    # 1. Count each elements occurences:
    for i in arr:
        cnt[i] += 1
    # 2. Generate sorted array:
    j = 0
    for i in range(0, MAX_VALUE):
        while(cnt[i]):
            arr[j] = i
            j += 1
            cnt[i] -= 1
    return arr

def main():
    arr = [int(x) for x in input().split()]
    counting_sort(arr,MAX_VALUE=10)
    print(arr)
    
if __name__ == "__main__":
    main()
